src/RemoveAtomicMutexLocks.cpp

```diff
@@ -0,0 +1,102 @@
+#include "RemoveAtomicMutexLocks.h"
+#include "IRVisitor.h"
+#include "IRMutator.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::string;
+using std::vector;
+
+namespace {
+
+class FindAtomicTupleProvide : public IRVisitor {
+    using IRVisitor::visit;
+
+    void visit(const Variable *op) override {
+        if (current_provide_name != "" &&
+                op->name == current_provide_name + ".value") {
```

Comment:
Finding variables by name like this is very fragile. Is there an alternative?

Reply:
Should be fixed. Now we search for let bindings variables that references the Provide nodes inside the same Atomic node.